src/Lean/Elab/PreDefinition/Structural.lean

```diff
@@ -354,6 +358,59 @@ private partial def replaceRecApps (recFnName : Name) (recArgInfo : RecArgInfo)
     | e => ensureNoRecFn recFnName e
   loop below e
 
+private partial def replaceIndPredRecApps (recFnName : Name) (recArgInfo : RecArgInfo) (motive : Expr) (e : Expr) : M Expr := do
+  let maxDepth := IndPredBelow.maxBackwardChainingDepth.get $ ←getOptions
+  let rec loop (e : Expr) : M Expr := do
+    match e with
+    | Expr.lam n d b c =>
+      withLocalDecl n c.binderInfo (← loop d) fun x => do
+        mkLambdaFVars #[x] (← loop (b.instantiate1 x))
+    | Expr.forallE n d b c =>
+      withLocalDecl n c.binderInfo (← loop d) fun x => do
+        mkForallFVars #[x] (← loop (b.instantiate1 x))
+    | Expr.letE n type val body _ =>
+      withLetDecl n (← loop type) (← loop val) fun x => do
+        mkLetFVars #[x] (← loop (body.instantiate1 x))
+    | Expr.mdata d e _   => return mkMData d (← loop e)
+    | Expr.proj n i e _  => return mkProj n i (← loop e)
+    | Expr.app _ _ _ =>
+      let processApp (e : Expr) : M Expr := do
+        let lCtx ← getLCtx
+        e.withApp fun f args => do
+          if f.isConstOf recFnName then
+            let ty ← inferType e
+            let main ← mkFreshExprSyntheticOpaqueMVar ty 
+            if ←IndPredBelow.backwardsChaining main.mvarId! maxDepth then
+              main
+            else
+              throwError "could not solve using backwards chaining {←Meta.ppGoal main.mvarId!}"
```

Comment:
We don't eagerly pretty print on error messages. We should use `{MessageData.ofGoal main.mvarId!}` instead of `{←Meta.ppGoal main.mvarId!}`.

Reply:
I remembered that but couldn't figure out how, thx.